In this assignment we shall use the Pandas concepts we have learnt to answer some questions about the Adults Data form the UCI ML Data Repository

In [1]:
#Importing Data and Libraries
import numpy as np
import pandas as pd

pd.set_option("display.max.columns", 100)
# to draw pictures in jupyter notebook
%matplotlib inline
# we don't like warnings
# you can comment the following 2 lines if you'd like to
import warnings

import matplotlib.pyplot as plt
import seaborn as sns

warnings.filterwarnings("ignore")


In [ ]:
DATA_URL = "https://raw.githubusercontent.com/Yorko/mlcourse.ai/main/data/"
data = pd.read_csv(DATA_URL + "adult.data.csv")

age         workclass  fnlwgt  education  education-num  \
0   39         State-gov   77516  Bachelors             13   
1   50  Self-emp-not-inc   83311  Bachelors             13   
2   38           Private  215646    HS-grad              9   
3   53           Private  234721       11th              7   
4   28           Private  338409  Bachelors             13   

       marital-status         occupation   relationship   race     sex  \
0       Never-married       Adm-clerical  Not-in-family  White    Male   
1  Married-civ-spouse    Exec-managerial        Husband  White    Male   
2            Divorced  Handlers-cleaners  Not-in-family  White    Male   
3  Married-civ-spouse  Handlers-cleaners        Husband  Black    Male   
4  Married-civ-spouse     Prof-specialty           Wife  Black  Female   

   capital-gain  capital-loss  hours-per-week native-country salary  
0          2174             0              40  United-States  <=50K  
1             0             0              13  United-States  <=50K  
2             0             0              40  United-States  <=50K  
3             0             0              40  United-States  <=50K  
4             0             0              40           Cuba  <=50K

Q1
How many men and women (sex feature) are represented in this dataset?

In [11]:
data["sex"].value_counts()

sex
Male      21790
Female    10771
Name: count, dtype: int64

Answer
Count of Men and Women in the Dateset
Men = 21790
Female =10771

Q2
What is the average age (age feature) of women?

In [14]:
data[data["sex"]=="Female"]["age"].mean()

np.float64(36.85823043357163)

Answer
The average age of all females in the dataset is 36.86 to 2 d.p

Q3
What is the percentage of German citizens (native-country feature)?

In [32]:
data["native-country"].value_counts(normalize=True)["Germany"]*100

np.float64(0.42074874850281013)

Answer
0.42% to 2 d.p

Q4-5. What are the mean and standard deviation of age for those who earn more than 50K per year (salary feature) and those who earn less than 50K per year?

In [39]:
data[data["salary"]=="<=50K"]["age"].aggregate(["mean", "std"])

mean    36.783738
std     14.020088
Name: age, dtype: float64

The mean and standard deviation for the age of people earning less than <=50K is 36.78 and 14.02 respectively

Q6 Is it true that people who earn more than 50K have at least high school education? (education – Bachelors, Prof-school, Assoc-acdm, Assoc-voc, Masters or Doctorate feature)

In [51]:
set(data[data["salary"]==">50K"]["education"].unique()) == set(["Bachelors", "Prof-school", "Assoc-acdm", "Assoc-voc", "Masters", "Doctorate"])

False

The claim is false , as the set of education features of people earning more than 50K does not equate to the Set given in the question

Q7.Display age statistics for each race (race feature) and each gender (sex feature). Use groupby() and describe(). Find the maximum age of men of Amer-Indian-Eskimo race.

In [55]:
columns = ["age"]
data.groupby(["race"])[columns].describe()


age                                                \
                      count       mean        std   min   25%   50%   75%   
race                                                                        
Amer-Indian-Eskimo    311.0  37.173633  12.447130  17.0  28.0  35.0  45.5   
Asian-Pac-Islander   1039.0  37.746872  12.825133  17.0  28.0  36.0  45.0   
Black                3124.0  37.767926  12.759290  17.0  28.0  36.0  46.0   
Other                 271.0  33.457565  11.538865  17.0  25.0  31.0  41.0   
White               27816.0  38.769881  13.782306  17.0  28.0  37.0  48.0   

                          
                     max  
race                      
Amer-Indian-Eskimo  82.0  
Asian-Pac-Islander  90.0  
Black               90.0  
Other               77.0  
White               90.0

In [58]:
columns = ["age"]
data.groupby(["sex"])[columns].describe()


age                                                    
          count       mean        std   min   25%   50%   75%   max
sex                                                                
Female  10771.0  36.858230  14.013697  17.0  25.0  35.0  46.0  90.0
Male    21790.0  39.433547  13.370630  17.0  29.0  38.0  48.0  90.0

The maximum age of the race Amer-Indian-Eskimo is 82

Q8.Among whom is the proportion of those who earn a lot (>50K) greater: married or single men (marital-status feature)? Consider as married those who have a marital-status starting with Married (Married-civ-spouse, Married-spouse-absent or Married-AF-spouse), the rest are considered bachelors.

In [68]:
data["Married"] = data["marital-status"].apply(lambda x : x[0] == "M")
data[data["salary"]==">50K"]["Married"].value_counts(normalize=True)

Married
True     0.859074
False    0.140926
Name: proportion, dtype: float64

People who hare married have make up the majority of people who earn more than 50K

Q9.What is the maximum number of hours a person works per week (hours-per-week feature)? How many people work such a number of hours, and what is the percentage of those who earn a lot (>50K) among them?

In [74]:
data["hours-per-week"].max() #99
data[data["hours-per-week"]==99].shape[0] #85
data[data["hours-per-week"]==99]["salary"].value_counts(normalize=True) 

salary
<=50K    0.705882
>50K     0.294118
Name: proportion, dtype: float64

The proportion of people earning >50K in the group of people who work 99 Hours is 0.705

In [94]:
data.columns
columns = ["hours-per-week"]
data.groupby(["native-country"])[columns].mean().loc["Japan"]



hours-per-week    43.693548
Name: Japan, dtype: float64

The average Hours per week for all the people living in Japan is 43.6935